src/main/java/org/logstash/dissect/JavaDissectorLibrary.java

```diff
@@ -1,281 +1,256 @@
+            }
+            plugin = (RubyObject) args[1];
+            pluginMetaClass = plugin.getMetaClass();
+            conversions = ConvertPair.createArrayFromHash((RubyHash) args[2]);
+            for (final ConvertPair convertPair : conversions) {
+                if (convertPair.converter().isInvalid()) {
+                    final RubyClass klass = ruby.getModule("LogStash").getClass("ConvertDatatypeFormatError");
+                    final String errorMessage = String.format("Dissector datatype conversion, datatype not supported: %s", convertPair.type());
+                    throw new RaiseException(ruby, klass, errorMessage, true);
                 }
-            });
+            }
+            runMatched = args[3] == null || args[3].isTrue();
+            failureTags = fetchFailureTags(ctx);
             return ctx.nil;
         }
 
         // def dissect(event, self)
-        @JRubyMethod(name = "dissect", required = 2)
-        public IRubyObject dissect(ThreadContext ctx, IRubyObject arg1, IRubyObject arg2) {
-            RubyEvent re = (RubyEvent) arg1;
-            if (re.ruby_cancelled(ctx).isTrue()) {
+        @JRubyMethod(name = "dissect", required = 1)
+        public final IRubyObject dissect(final ThreadContext ctx, final IRubyObject arg1) {
+            final JrubyEventExtLibrary.RubyEvent rubyEvent = (JrubyEventExtLibrary.RubyEvent) arg1;
+            if (rubyEvent.ruby_cancelled(ctx).isTrue()) {
                 return ctx.nil;
             }
-            Event e = re.getEvent();
-            RubyObject plugin = (RubyObject) arg2;
-            RubyObject logger = (RubyObject) getLoggerObject(ctx, plugin);
+            final Event event = rubyEvent.getEvent();
             try {
-                if (logLevelEnabled(ctx, logger, "debug")) {
-                    logDebug(ctx, logger, "Event before dissection", buildDebugEventMap(ctx, re));
-                }
-                for (Map.Entry<RubyString, Dissector> entry : dissectors.entrySet()) {
-                    RubyString key = entry.getKey();
-                    if (e.includes(key.toString())) {
-                        // use ruby event here because we want the bytelist bytes
-                        // from the ruby extract without converting to Java
-                        RubyString src = re.ruby_get_field(ctx, key).asString();
-                        entry.getValue().dissect(src.getBytes(), e);
-                    } else {
-                        HashMap<IRubyObject, IRubyObject> map = new HashMap<>(2);
-                        map.put(rubyString(ctx, "key"), key);
-                        map.put(rubyString(ctx, "event"), re.ruby_to_hash(ctx));
-                        logWarn(ctx, logger, "Dissector mapping, key not found in event", map);
-                    }
+                LOGGER.debug("Event before dissection", buildLoggerEventMap(event));
```

Comment:
in a tight loop, we would create a lot of unecessary map everytime the logger is called, we should add a simple boolean everywhere we do this.
```java
if (LOGGER.isDebugEnabled()) {
 LOGGER.debug("Event before dissection", buildLoggerEventMap(event));
}
```

Reply:
in log4j2, I believe we can use a closure.

Reply:
I tried the lambda approach but the message field value was missing?
Going with the `if (enabled) { log }` approach.